dataflow/flex-templates/streaming_beam/streaming_beam.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python
+#
+# Copyright 2020 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""An Apache Beam streaming pipeline example.
+
+It reads JSON encoded messages from Pub/Sub, transforms the message data and
+writes the results to BigQuery.
+"""
+
+from __future__ import absolute_import
+
+import argparse
+import json
+import logging
+import time
+
+import apache_beam as beam
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+from apache_beam.options.pipeline_options import StandardOptions
+import apache_beam.transforms.window as window
+
+
+# Defines the BigQuery schema for the output table.
+SCHEMA = ('url:STRING,'
+          'num_reviews:INTEGER,'
+          'score:FLOAT64,'
+          'first_date:TIMESTAMP,'
+          'last_date:TIMESTAMP')
+
+
+class ParseJsonMessageDoFn(beam.DoFn):
+  """Parse the json encoded text message into a dict."""
+
+  def __init__(self):
+    beam.DoFn.__init__(self)
+
+  def process(self, element):
+    """Returns an iterator over the tuple: (message's url, dict).
+
+    Args:
+      element: the element json message to be processed.
+    Returns:
+      The tuple of url string and the dict.
+      The dict contain the keys: 'url', 'review', 'score' and 'processing_time'.
+    """
+    row = json.loads(element)
+    row['score'] = 0.0
+    if row['review'] == 'positive':
+      row['score'] = 1.0
+    row['processing_time'] = int(time.time())
+    return [(row['url'], row)]
+
+
+def run(argv=None, save_main_session=True):
+  """Build and run the pipeline."""
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      '--output_table',
+      help=(
+          'Output BigQuery table for results specified as: '
+          'PROJECT:DATASET.TABLE or DATASET.TABLE.'))
+
+  group = parser.add_mutually_exclusive_group(required=True)
+  group.add_argument(
+      '--input_topic',
+      help=(
+          'Input PubSub topic of the form '
+          '"projects/<PROJECT>/topics/<TOPIC>".'))
+  group.add_argument(
+      '--input_subscription',
+      help=(
+          'Input PubSub subscription of the form '
+          '"projects/<PROJECT>/subscriptions/<SUBSCRIPTION>."'))
+  known_args, pipeline_args = parser.parse_known_args(argv)
+
+  pipeline_options = PipelineOptions(pipeline_args)
```

Comment:
We always want the `save_main_session` option so I would hardcode it to `True`. You can also initialize those flags directly in the `PipelineOptions` constructor instead of having to cast to other subtypes directly. I would also be okay to rename `pipeline_options` to `options` since there is nothing else to disambiguate from.

```py
options = PipelineOptions(
  args,
  save_main_session=True,
  streaming=True)
```

Reply:
Done.